javascript/sdk/src/repl.ts

```diff
@@ -41,145 +40,101 @@ export type MessageFromServer =
-
-      case 'output': {
-        if (this.state.type !== 'waiting_for_result') {
-          console.warn('Unexpected message in state', this.state, 'with message', message)
-          return
-        }
-
-        if (message.instruction_id !== this.state.instruction_id) {
-          console.warn('Unexpected instruction id', this.state, 'with message', message)
-          return
-        }
-
-        this.state.outputCallback(message.chunk)
-        break
-      }
-
-      case 'result': {
-        if (this.state.type !== 'waiting_for_result') {
-          console.warn('Unexpected message in state', this.state, 'with message', message)
-          return
-        }
-
-        if (message.instruction_id !== this.state.instruction_id) {
-          console.warn('Unexpected instruction id', this.state, 'with message', message)
-          return
-        }
-
-        this.state.resultCallback(message.result)
-        this.state = { type: 'idle' }
-        break
-      }
-    }
-  }
-
   private send(message: MessageToServer) {
     this.ws.send(JSON.stringify(message))
   }
 
-  async exec(code: string, options: ExecOptions = {}): Promise<ReplExecResult> {
+  exec(code: string, options: ExecOptions = {}): ReplExecResult {
     const request_id = this.nextRequestId++
     const instruction = {
       code,
       max_runtime_ms: options.timeoutSeconds,
       timeout_seconds: options.timeoutSeconds,
     }
-    const instructionSeq = await new Promise<number>((resolve) => {
-      this.state = {
-        type: 'waiting_for_instruction_seq',
-        request_id,
-        callback: resolve,
-      }
-      this.send({ type: 'exec', instruction, request_id })
-    })
 
-    const result = new ReplExecResult()
-    this.state = {
-      type: 'waiting_for_result',
-      instruction_id: instructionSeq,
-      resultCallback: (execResult) => result.setResult(execResult),
-      outputCallback: (output) => result.addOutput(output),
-    }
-    return result
+    this.send({ type: 'exec', instruction, request_id })
+    this.listener = new EventTarget()
+    return new ReplExecResult(this.listener)
   }
 }
 
 export class ReplExecResult {
-  private resultPromise: Promise<ExecResponse>
-  private output: Stream<StandardOutput> = new Stream()
-  private resultResolver?: (result: ExecResponse) => void
-
-  constructor() {
-    this.resultPromise = new Promise<ExecResponse>((resolve) => {
-      this.resultResolver = resolve
-    })
+  #listener: EventTarget
+  #done = false
+  #resolve: (response: ExecResponse) => void
+  #reject: (reason: any) => void
+
+  #buffer: StandardOutput[] = []
+  #advance: (() => void) | undefined = undefined
+
+  result: Promise<ExecResponse>
+  output: AsyncIterable<StandardOutput, undefined>
+
+  constructor(listener: EventTarget) {
+    this.#listener = listener
+    this.#listener.addEventListener('exec_received', this)
+    this.#listener.addEventListener('output', this)
+    this.#listener.addEventListener('result', this)
```

Comment:
since this is the only place we are subscribing, should we just throw these all into one `msg` event type?